packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesUtils.kt

```diff
@@ -160,58 +160,67 @@ object UpdatesUtils {
-            // if it has an instance first.
-            if (host.hasInstance()) {
-              reactContext = host.reactInstanceManager.currentReactContext
-              if (reactContext != null) {
-                break
-              }
-            }
-            Thread.sleep(1000)
-          }
-          if (reactContext != null) {
-            val emitter = reactContext.getJSModule(
-              DeviceEventManagerModule.RCTDeviceEventEmitter::class.java
-            )
-            if (emitter != null) {
-              var eventParams = params
-              if (eventParams == null) {
-                eventParams = Arguments.createMap()
-              }
-              eventParams!!.putString("type", eventType)
-              logger.info("Emitted event: name = $eventName, type = $eventType")
-              emitter.emit(eventName, eventParams)
-              return@execute
-            }
-          }
-          logger.error("Could not emit $eventName $eventType event; no event emitter was found.", UpdatesErrorCode.JSRuntimeError)
-        } catch (e: Exception) {
-          logger.error("Could not emit $eventName $eventType event; no react context was found.", UpdatesErrorCode.JSRuntimeError)
-        }
-      }
-    } else {
-      logger.error(
-        "Could not emit $eventType event; UpdatesController was not initialized with an instance of ReactApplication.",
-        UpdatesErrorCode.Unknown
-      )
+    val eventParams = params ?: Arguments.createMap()
+    eventParams.putString("type", eventType)
+
+    if (!shouldEmitJsEvents) {
+      eventsToSendToJS.add(Pair(eventName, eventParams))
+      logger.error("Could not emit $eventName $eventType event; no subscribers registered.", UpdatesErrorCode.JSRuntimeError)
+      return
+    }
+    val appContext = weakAppContext?.get() ?: run {
+      eventsToSendToJS.add(Pair(eventName, eventParams))
+      logger.error("Could not emit $eventName $eventType event; no app context was found.", UpdatesErrorCode.JSRuntimeError)
+      return
+    }
+    val updatesModule = appContext.registry.getModule("ExpoUpdates") ?: run {
+      eventsToSendToJS.add(Pair(eventName, eventParams))
+      logger.error("Could not emit $eventName $eventType event; no ExpoUpdates module was found.", UpdatesErrorCode.JSRuntimeError)
+      return
+    }
+    val eventEmitter = appContext.eventEmitter(updatesModule) ?: run {
+      eventsToSendToJS.add(Pair(eventName, eventParams))
+      logger.error("Could not emit $eventName $eventType event; no event emitter was found.", UpdatesErrorCode.JSRuntimeError)
+      return
+    }
+
+    logger.info("Emitted event: name = $eventName, type = $eventType")
+    eventEmitter.emit(eventName, eventParams)
+  }
+
+  fun sendQueuedEventsToAppContext(
+    shouldEmitJsEvents: Boolean,
+    weakAppContext: WeakReference<AppContext>?,
+    logger: UpdatesLogger
+  ) {
+    if (!shouldEmitJsEvents) {
```

Comment:
should we have the same `logger.error` calls in these guard clauses as we do in the `sendEventToAppContext` guard clauses?

Reply:
for sending "queued" events, it looks like we didn't log when things are not ready: https://github.com/expo/expo/blob/23929f09c1b9c7f74390d4eba394a4a7a95585eb/packages/expo-updates/ios/EXUpdates/EnabledAppController.swift#L225-L233. i could add log for both ios and android though. wdyt?

Reply:
makes sense to me. thanks for sharing that!